Clients/Xamarin.Interactive.Client.Mac/ViewInspectorMainViewController.cs

```diff
@@ -1,165 +1,82 @@
-                    if (!supportedHierarchies.Contains (selectedHierarchy))
-                        selectedHierarchy = supportedHierarchies [0];
-
-                    remoteView = await Session.Agent.Api.GetVisualTreeAsync (
-                        selectedHierarchy,
-                        captureViews: true);
-                } else {
-                    supportedHierarchies = null;
-                    selectedHierarchy = null;
-                }
-            } catch (Exception e) {
-                e.ToUserPresentable (Catalog.GetString ("Error trying to inspect remote view"))
-                    .Present (this);
-            }
-
-            SelectView (remoteView, withinExistingTree: false, setSelectedView: false);
-
-            UpdateHierarchiesList (supportedHierarchies, selectedHierarchy);
-
-            isVisualTreeRefreshing = false;
-
-            View?.Window?.Toolbar?.ValidateVisibleItems ();
+            base.OnSessionAvailable ();
+            model = new MacViewInspector (this, Session);
         }
 
-        public void UpdateSelectedHierarchy (string selectedHierarchy)
-        {
-            this.selectedHierarchy = selectedHierarchy;
-            RefreshVisualTree (this);
-        }
+        protected override void OnAgentFeaturesUpdated ()
+            => RefreshVisualTree (this);
 
-        public void UpdateHierarchiesList (string [] supportedHierarchies, string selectedHierarchy)
-        {
-            ChildViewControllers.OfType<ViewHierarchyViewController> ().ForEach (vc => {
-                vc.UpdateSupportedHierarchies (supportedHierarchies, selectedHierarchy);
-            });
-        }
+        public void UpdateSelectedHierarchy (string selectedHierarchy)
+            => model.SelectedHierarchy = selectedHierarchy;
 
         public void SelectView (InspectView view)
             => SelectView (view, false, true);
 
-        void SelectView (InspectView view, bool withinExistingTree, bool setSelectedView)
-        {
-            if (!Session.Agent.IsConnected)
-                return;
+        void SelectView (InspectView view, bool withinExistingTree = false, bool setSelectedView = true)
+            => model.SelectView (view, withinExistingTree, setSelectedView);
 
-            if (setSelectedView
-                && !string.IsNullOrEmpty (view?.PublicCSharpType)
-                && Session.SessionKind == ClientSessionKind.LiveInspection) {
-                var code = $"var selectedView = GetObject<{view.PublicCSharpType}> (0x{view.Handle:x})";
-                Session.WorkbookPageViewModel.EvaluateAsync (code).Forget ();
+        static T TagToEnum<T> (nint tag)
+        {
+            // Keep these synced to the tags for the 2D/3D menu items in the Main storyboard
+            object result = default (T);
+            switch (result) {
+            case DisplayMode _:
+                return (T)(object)(DisplayMode)(uint)(tag - 4);
```

Comment:
Wow. You can probably use `Enum.ToObject` here though: `(T)Enum.ToObject (typeof (T), (uint)(tag - 4))`. If the values were in sync, you could just condense the entire method to `Enum.ToObject`.